packages/devtools_app/lib/src/eval_on_dart_library.dart

```diff
@@ -156,25 +164,176 @@ class EvalOnDartLibrary {
   }
 
-  Future<Class> getClass(ClassRef instance, ObjectGroup isAlive) {
+  Future<Class> getClass(ClassRef instance, Disposable isAlive) {
     return getObjHelper(instance, isAlive);
   }
 
-  Future<Func> getFunc(FuncRef instance, ObjectGroup isAlive) {
+  Future<Func> getFunc(FuncRef instance, Disposable isAlive) {
     return getObjHelper(instance, isAlive);
   }
 
   Future<Instance> getInstance(
     FutureOr<InstanceRef> instanceRefFuture,
-    ObjectGroup isAlive,
+    Disposable isAlive,
   ) async {
     return await getObjHelper(await instanceRefFuture, isAlive);
   }
 
+  Future<String> getInstanceHashCode(
+    InstanceRef instance, {
+    @required Disposable isAlive,
+  }) async {
+    // TODO(rrousselGit) figure out why sometimes this returns null
+    final hash = await evalInstance(
+      'instance.hashCode.toUnsigned(20).toRadixString(16).padLeft(5, "0")',
+      isAlive: isAlive,
+      scope: {'instance': instance.id},
+    );
+
+    return hash.valueAsString;
+  }
+
+  Future<Instance> evalInstance(
+    String expression, {
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    final ref = await safeEval(expression, isAlive: isAlive, scope: scope);
+
+    if (ref == null) return null;
+
+    return getInstance(ref, isAlive);
+  }
+
+  static int _nextId = 0;
+
+  /// a [safeEval] variant that does not complete until the Future evaluated completes
+  ///
+  /// The expression **must** return a [FutureOr].
+  ///
+  /// Works only if the evaluated library includes the following code:
+  ///
+  /// ```dart
+  /// Future<T> $await<T>(Future<T> future, String id) async {
+  ///   try {
+  ///     return await future;
+  ///   } finally {
+  ///     postEvent('future_completed', {'id': id});
+  ///   }
+  /// }
+  /// ```
+  Future<InstanceRef> awaitEval(
+    String expression, {
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    final id = _nextId++;
+
+    // start awaiting the event before starting the evaluation, in case the
+    // event is received before the eval function completes.
+    final future = serviceManager.service.onExtensionEvent.firstWhere((event) {
+      return event.extensionKind == 'future_completed' &&
+          event.extensionData.data['id'] == id;
+    });
+
+    final result = await safeEval(
+      '''
+() async { try { return await $expression; } finally { postEvent('future_completed', {'id': $id}); } }()
+''',
+      isAlive: isAlive,
+      scope: scope,
+    );
+
+    await future;
+
+    return result;
+  }
+
+  /// An [eval] that does not return `null` when a [Sentinel] or an error occurs
+  Future<InstanceRef> safeEval(
+    String expression, {
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    try {
+      final result = await addRequest(
+        isAlive,
+        () => _safeEval(expression, scope: scope),
+      );
+      if (isAlive.disposed) {
+        throw StateError(
```

Comment:
why is this an error? shouldn't it just mean that the request has been cancelled and we should return a null InstanceRef?